// @ngneat/spectator/esm2015/lib/spectator-routing/router-link-stub.js
import { Directive, HostListener, Input, NgModule } from '@angular/core';
import { RouterLink } from '@angular/router';
// tslint:disable
export class RouterLinkDirectiveStub {
    constructor() {
        this.navigatedTo = null;
    }
    onClick() {
        this.navigatedTo = this.routerLink;
        return true;
    }
}
RouterLinkDirectiveStub.decorators = [
    { type: Directive, args: [{
                selector: '[routerLink]',
                providers: [
                    {
                        provide: RouterLink,
                        useExisting: RouterLinkDirectiveStub
                    }
                ]
            },] }
];
RouterLinkDirectiveStub.propDecorators = {
    routerLink: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click',] }]
};
/*
  This is an unused module to resolve the ng build error:
    'Cannot determine the module for class RouterLinkDirectiveStub'

  Reference: https://github.com/angular/issues/13590
*/
export class RouterLinkDirectiveStubModule {
}
RouterLinkDirectiveStubModule.decorators = [
    { type: NgModule, args: [{
                declarations: [RouterLinkDirectiveStub]
            },] }
];
